sdk/identity/identity/src/credentials/managedIdentityCredential/msalMsiProvider.ts

```diff
@@ -0,0 +1,271 @@
+        // AKS pod identity - MSAL does not implement the token exchange flow.
+        // IMDS Endpoint probing - MSAL does not do any probing before trying to get a token.
+        // As a DefaultAzureCredential optimization we probe the IMDS endpoint with a short timeout and no retries before actually trying to get a token
+        // We will continue to implement these features in the Identity library.
+
+        const identitySource = this.managedIdentityApp.getManagedIdentitySource();
+        const isImdsMsi = identitySource === "DefaultToImds" || identitySource === "Imds"; // Neither actually checks that IMDS endpoint is available, just that it's the source the MSAL _would_ try to use.
+
+        if (isTokenExchangeMsi) {
+          // In the AKS scenario we will use the existing tokenExchangeMsi indefinitely.
+          logger.getToken.info("Using the token exchange managed identity.");
+          const result = await tokenExchangeMsi.getToken({
+            scopes,
+            clientId: this.clientId,
+            identityClient: this.identityClient,
+            retryConfig: this.msiRetryConfig,
+            resourceId: this.resourceId,
+          });
+
+          if (result === null) {
+            throw new CredentialUnavailableError(
+              "The managed identity endpoint was reached, yet no tokens were received.",
+            );
+          }
+
+          return result;
+        } else if (isImdsMsi) {
+          // In the IMDS scenario we will probe the IMDS endpoint to ensure it's available before trying to get a token.
+          // If the IMDS endpoint is not available and this is the source that MSAL will use, we will fail-fast with an error that tells DAC to move to the next credential.
+          logger.getToken.info("Using the IMDS endpoint to probe for availability.");
+          const isAvailable = await imdsMsi.isAvailable({
+            scopes,
+            clientId: this.clientId,
+            getTokenOptions: options,
+            identityClient: this.isAvailableIdentityClient,
+            resourceId: this.resourceId,
+          });
+
+          if (!isAvailable) {
+            throw new CredentialUnavailableError(
+              `ManagedIdentityCredential: The managed identity endpoint is not available.`,
+            );
+          }
+        }
+
+        // If we got this far, it means:
+        // - This is not a tokenExchangeMsi,
+        // - We already probed for IMDS endpoint availability and failed-fast if it's unreachable.
+        // We can proceed normally by calling MSAL for a token.
+        logger.getToken.info("Calling into MSAL for managed identity token.");
+        const token = await this.managedIdentityApp.acquireToken({
+          resource,
+        });
+
+        this.ensureValidMsalToken(scopes, token, options);
+        logger.getToken.info(formatSuccess(scopes));
+
+        return {
+          expiresOnTimestamp: token.expiresOn.getTime(),
+          token: token.accessToken,
```

Comment:
is `refreshOn` included here also?

Reply:
Not in this PR, but Karishma's PR https://github.com/Azure/azure-sdk-for-js/pull/30402 will include support for refreshOn in both managed identity flows (legacy / existing + msal-based / this PR's changes)

Reply:
It's not included as part of this PR. Here's the PR that I am working towards for this feature - https://github.com/Azure/azure-sdk-for-js/pull/30402